File: src/utils/favorite.js
/**
 * 课程收藏 API
 */

import { request } from './cloud-db.js'

/**
 * 获取用户收藏列表
 */
export async function getFavorites(userId) {
  try {
    const res = await request('GET', '/api/favorites', { user_id: userId })
    return {
      ok: true,
      list: res?.list || [],
      total: res?.total || 0
    }
  } catch (error) {
    console.error('获取收藏列表失败:', error)
    return { ok: false, list: [], total: 0 }
  }
}

/**
 * 添加收藏
 */
export async function addFavorite(userId, courseId, isExternal = true) {
  try {
    const res = await request('POST', '/api/favorites', {
      user_id: userId,  // 添加 user_id 到请求体
      course_id: courseId,
      is_external: isExternal
    })
    return { ok: true, id: res.id }
  } catch (error) {
    console.error('添加收藏失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 取消收藏
 */
export async function removeFavorite(favoriteId) {
  try {
    await request('DELETE', `/api/favorites/${favoriteId}`)
    return { ok: true }
  } catch (error) {
    console.error('取消收藏失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 根据课程ID取消收藏
 */
export async function removeFavoriteByCourse(userId, courseId, isExternal = true) {
  try {
    await request('DELETE', `/api/favorites/course?user_id=${userId}&course_id=${courseId}&is_external=${isExternal}`)
    return { ok: true }
  } catch (error) {
    console.error('取消收藏失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 检查是否已收藏
 */
export async function checkFavorite(userId, courseId, isExternal = true) {
  try {
    const res = await request('GET', '/api/favorites/check', {
      user_id: userId,
      course_id: courseId,
      is_external: isExternal
    })
    return {
      ok: true,
      isFavorited: res?.is_favorited || false,
      favoriteId: res?.favorite_id || null
    }
  } catch (error) {
    console.error('检查收藏状态失败:', error)
    return { ok: false, isFavorited: false, favoriteId: null }
  }
}

File: backend/app/routers/favorite.py
"""
课程收藏路由
"""
import time
from typing import Optional
from fastapi import APIRouter, HTTPException, Query, Body, Header
from pydantic import BaseModel

from app.models.favorite import CourseFavorite
from app.models.external_course import ExternalCourse

router = APIRouter(prefix="/api/favorites", tags=["课程收藏"])


class FavoriteResponse(BaseModel):
    id: int
    course_id: int
    is_external: bool
    created_at: int
    course: dict = None


def _get_user_id(x_user_id: str = Header(None), user_id: int = Query(None)):
    """从请求头或查询参数获取用户ID"""
    if x_user_id:
        return int(x_user_id)
    if user_id:
        return user_id
    return None


@router.get("")
async def get_favorites(x_user_id: str = Header(None), user_id: int = Query(None)):
    """获取用户的所有收藏"""
    uid = _get_user_id(x_user_id, user_id)
    if not uid:
        return {"ok": True, "list": [], "total": 0}
    
    favorites = await CourseFavorite.all().filter(user_id=uid).order_by("-created_at")

    result = []
    for fav in favorites:
        course_info = None
        if fav.is_external:
            course = await ExternalCourse.filter(id=fav.course_id).first()
            if course:
                course_info = {
                    "id": course.id,
                    "title": course.title,
                    "cover": course.cover,
                    "link": course.link,
                    "description": course.description,
                    "category": course.category,
                }

        result.append({
            "id": fav.id,
            "course_id": fav.course_id,
            "is_external": fav.is_external,
            "created_at": fav.created_at,
            "course": course_info
        })

    return {"ok": True, "list": result, "total": len(result)}


class AddFavoriteBody(BaseModel):
    user_id: Optional[int] = None
    course_id: int
    is_external: bool = True


@router.post("")
async def add_favorite(
    body: AddFavoriteBody,
    x_user_id: str = Header(None)
):
    """添加收藏"""
    # 优先使用 header 中的用户ID
    if x_user_id:
        user_id = int(x_user_id)
    elif body.user_id:
        user_id = body.user_id
    else:
        raise HTTPException(status_code=401, detail="请先登录")
    
    course_id = body.course_id
    is_external = body.is_external

    # 检查是否已收藏
    existing = await CourseFavorite.filter(
        user_id=user_id,
        course_id=course_id,
        is_external=is_external
    ).first()

    if existing:
        return {"ok": True, "id": existing.id, "message": "已收藏"}

    now = int(time.time() * 1000)
    favorite = await CourseFavorite.create(
        user_id=user_id,
        course_id=course_id,
        is_external=is_external,
        created_at=now
    )

    return {"ok": True, "id": favorite.id}


@router.delete("/course")
async def remove_favorite_by_course(
    x_user_id: str = Header(None),
    user_id: int = Query(None),
    course_id: int = Query(...),
    is_external: str = Query("true")
):
    """根据课程ID取消收藏"""
    uid = _get_user_id(x_user_id, user_id)
    if not uid:
        raise HTTPException(status_code=401, detail="请先登录")

    # 转换字符串到布尔值
    is_ext = is_external.lower() in ('true', '1', 'yes') if isinstance(is_external, str) else bool(is_external)

    favorite = await CourseFavorite.filter(
        user_id=uid,
        course_id=course_id,
        is_external=is_ext
    ).first()

    if not favorite:
        raise HTTPException(status_code=404, detail="收藏不存在")

    await favorite.delete()
    return {"ok": True}


@router.delete("/{favorite_id}")
async def remove_favorite(favorite_id: int):
    """取消收藏"""
    favorite = await CourseFavorite.filter(id=favorite_id).first()
    if not favorite:
        raise HTTPException(status_code=404, detail="收藏不存在")

    await favorite.delete()
    return {"ok": True}


@router.get("/check")
async def check_favorite(
    x_user_id: str = Header(None),
    user_id: int = Query(None),
    course_id: int = Query(...),
    is_external: bool = Query(True)
):
    """检查是否已收藏"""
    uid = _get_user_id(x_user_id, user_id)
    if not uid:
        return {"ok": True, "is_favorited": False, "favorite_id": None}
    
    favorite = await CourseFavorite.filter(
        user_id=uid,
        course_id=course_id,
        is_external=is_external
    ).first()

    return {"ok": True, "is_favorited": favorite is not None, "favorite_id": favorite.id if favorite else None}
